packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js

```diff
@@ -3,9 +3,17 @@ const debug = require(`debug`)(`gatsby:component-shadowing`)
 const fs = require(`fs`)
 const _ = require(`lodash`)
 
-const pathWithoutExtension = fullPath => {
-  const parsed = path.parse(fullPath)
-  return path.join(parsed.dir, parsed.name)
+// By default, a file can only be shadowed by a file of the same extension.
+// However, the following table determine additionnal shadowing extensions that
+// will be looked for, given the extension of the file being shadowed.
+// This list maybe extended by user (by customizing webpack's configuration), in
+// order to allow less common use cases (ie. allow css files being shadowed by
+// a scss file, or jpg files being shadowed by png...)
+const DEFAULT_ADDITIONNAL_SHADOW_EXTENSIONS = {
```

Comment:
While we like the idea of having an explicit map of shadow extensions, people rely on `extensions: getConfig().resolve.extensions` to exist, e.g. for shadowing MDX files or in (probably) some rare cases even different image formats.

Was this something that was required to make shadowing in yarn workspaces work or was this thought of an improvement from you?

Reply:
I absolutely understand what you are saying and I would have prefered not to introduce a breaking change here. This was not designed as a deliberate improvement (though it might indeed be one, if can make it easier to work with). I'm open to discussion if someone sees alternatives, but I haven't found any way yet that do not introduces even worst consequences. So let me explain the situation, in case someone gets an idea.

It is expected that theme shadowing allows a shadow file to be of a different file extension that the file being shadowed... But in Gatsby 2, this relied on the fact that an importer file would import the file without specifying the file extension. For example, importer.js would import `gatsby-theme-example/Component` (without file extension).

Note however that if importer.js actually imported a file using a specific file extension, then the theme shadowing plugin would never try to _remove_ that file extension (though it would still look up for that file name with every entries in `extensions` appended). For example, importing `gatsby-theme-example/Component.css` would look for shadow files named `Component.css`, `Component.css.js`, `Component.css.ts`... But not for `Component.js`, which would be unexpected and certainly inappropriate. Therefore, _this is certainly the behavior that we want to preserve._ As we will see, this is the difficult point.

Now, In Gatsby 2 / Webpack 4, this was possible because the Webpack's resolver plugin would get invoked at some point where `request.path` is `/full/and/canonical/path/to/gatsby-theme-example/src/Component`. Note that at this point, the name of the package has been replaced by the absolute and canonical path (i.e. no symlink) to the source directory of the theme, but the file itself still has no extension.

In Webpack 5, there is never such a moment where the package's path has been fully resolved, but the file extension is still left to its original value. Indeed, symlinks are now canonicalized at the very last step of resolution, leaving very opportunity for interception of the resolution process by the shadowing plugin. At this point, Webpack's resolution process has already determined the file extension, based on the content of the original package (i.e. files in the unshadowed theme's source directory). For example, if importer.js requested `gatsby-theme-example/Component`, the plugin will get called with something like `/full/and/canonical/path/to/gatsby-theme-example/src/Component.js`, and the plugin gets no cue that the file that was originally requested is `Component`, without extension.

So the question is: how can the plugin determine at that point that `Component.ts` (that is an example) is an acceptable substitution file for the file specified in `request.path` (i.e. `Component.js`)? I initially thought of removing the provided file extension from `request.path` and then lookup for all files matching that name + one of the extension listed in `extensions`.  So a requested path of `.../Component.js` would get back to `Component`, and then would look for shadow files such as `Component.js`, `Component.jsx`, `Component.ts` and `Component.tsx`.

But then, what if importer request a file that is not code, such as `Component.css` or `Component.jpeg`? Assuming the previous strategy, both gets trimmed down to `Component`, and the shadowing plugin would then happily return the content of `Component.js` instead of `Component.css` or `Component.jpeg`.

I thought of trimming only "known" extensions (that is, extensions listed in `extensions`), but this still is insufficient as it is sometime useful to have `css`, `jpeg` or other non-code file extensions defined in `extensions`.

So I came up with this. The idea is that a code file should only be shadowed by another code file, an image by an image, and so on. From the algorithm point of view, I think the current strategy makes things simple, strong and predictable.

From the user's point of view, however, the "additional shadow extensions" table is difficult to customize and it introduces a breaking change. I think it is that specific point that needs to be improved.

Reply:
One thing that might be worth considering is that it would be relatively easy to automatically construct this mapping table based on the content of `extensions` array if we had some type of "file extension dictionary". Such "dictionary" would for example contains:

```
js => code
jsx => code
ts => code
tsx => code
jpeg => image
png => image
css => stylesheet
sass => stylesheet
```

Then, based on the content of `extensions`, it is now a simple algorithm exercise to rebuild the mapping table described previously. Note that it would be acceptable for the dictionary to contains extensions that are not actually used in a project, making it possible for the plugin to provide a default dictionary containing a reasonable list of known common extensions. Only extensions specified in Webpack's `extensions` array would actually get included in the mapping table.

I can easily provide the algorithm to build up the mapping table, but I would need some guidance on where to store the dictionary so that it is customizable (and possibly how to propagate that information up to the webpack's plugin).

Reply:
> Now, In Gatsby 2 / Webpack 4, this was possible because the Webpack's resolver plugin would get invoked at some point where `request.path` is `/full/and/canonical/path/to/gatsby-theme-example/src/Component`. Note that at this point, the name of the package has been replaced by the absolute and canonical path (i.e. no symlink) to the source directory of the theme, but the file itself still has no extension.

Can we still use same hook we used (or even any earlier ones) and store that `request.path` without added extension (if this was how it was imported of course in source code) so that `before-resolved` hook (one that does actual work) have access to it? I don't exactly know if passing information between hooks is "forbidden" or not but this:
```
    resolver.hooks.relative.tapAsync(`wat`, (request, stack, callback) => {
      console.log(`relative`, request.path)
      request.hack = request.path
      return callback()
    })

    resolver
      .getHook(`before-resolved`)
      .tapAsync(
        `GatsbyThemeComponentShadowingResolverPlugin`,
        (request, stack, callback) => {
          console.log(`before-resolved`, {
            path: request.path,
            hack: request.hack,
          })
```
Resulted in
```
console.log packages/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:51
    relative /Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/index.js

  console.log packages/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:61
    before-resolved {
      path: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/index.js',
      hack: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/index.js'
    }

  console.log packages/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:51
    relative /Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/node_modules/theme-a/src/components

  console.log packages/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:61
    before-resolved {
      path: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/node_modules/theme-a/src/components.js',
      hack: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/node_modules/theme-a/src/components'
    }

  console.log packages/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:61
    before-resolved {
      path: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/src/theme-a/components.js',
      hack: '/Users/misiek/dev/gatsby/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/__tests__/fixtures/test-sites/basic-shadowing/node_modules/theme-a/src/components'
    }
```
Which made it seem possible to pass some meta information between hooks (tho I definitely don't like mutating `request` to achieve this - as name in my code snippet suggest - it seems very hacky and also relies on `request` being same object between hook invocations that seems like implementaion detail that shouldn't be relied upon).

> From the user's point of view, however, the "additional shadow extensions" table is difficult to customize and it introduces a breaking change. I think it is that specific point that needs to be improved.

I think for gatsby@3 we will need to maintain behaviour, but the shadow extensions matrix is definitely something we like (we just can't completely switch to it). I do think there is a way to introduce usage of such matrix without creating breaking changes by following something like this:
 - highest priority - same extension
 - if above doesn't exist - try to use extension matrix to prioritize known compatible (swappable) extensions
 - if match still not found - fallback to just list of resolvable extensions (to not introduce breaking change) - to me only problem with doing this fallback is the issue you outlined above that we currently in `before-resolved` don't know if file was imported with explicit extension or extension was inferred at some point by webpack - if we would have access to metadata what the import looked like we could use it to decide wether we use that fallback or not

Reply:
> Can we still use same hook we used (or even any earlier ones) and store that `request.path` without added extension (if this was how it was imported of course in source code) so that `before-resolved` hook (one that does actual work) have access to it? I don't exactly know if passing information between hooks is "forbidden" or not but this:

> Which made it seem possible to pass some meta information between hooks (tho I definitely don't like mutating request to achieve this - as name in my code snippet suggest - it seems very hacky and also relies on request being same object between hook invocations that seems like implementaion detail that shouldn't be relied upon).

I had initially thought of this option but had rejected it as too hacky... However, now that you bring it back on the table, I think it might be workable. I have examined a few default Webpack plugins, and this actually appears to be a strategy used internally. And though request objects themselves are never _reused_, new request objects always extends previous ones, so custom properties set early should still be available in later plugins.

I will implement something that integrate this as a fallback strategy, and come back with a new commit.